cirq-core/cirq/transformers/eject_z.py

```diff
@@ -0,0 +1,145 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Transformer pass that pushes Z gates later and later in the circuit."""
+
+from typing import Dict, Iterable, Optional, Tuple, TYPE_CHECKING
+from collections import defaultdict
+import numpy as np
+
+from cirq import ops, protocols
+from cirq.transformers import transformer_api, transformer_primitives
+from cirq.transformers.analytical_decompositions import single_qubit_decompositions
+
+if TYPE_CHECKING:
+    import cirq
+
+
+def _is_integer(n):
+    return np.isclose(n, np.round(n))
+
+
+def _is_swaplike(gate: 'cirq.Gate'):
+    if isinstance(gate, ops.SwapPowGate):
+        return gate.exponent == 1
+
+    if isinstance(gate, ops.ISwapPowGate):
+        return _is_integer((gate.exponent - 1) / 2)
+
+    if isinstance(gate, ops.FSimGate):
+        return _is_integer(gate.theta / np.pi - 1 / 2)
+
+    return False
+
+
+@transformer_api.transformer
+def eject_z(
+    circuit: 'cirq.AbstractCircuit',
+    *,
+    context: Optional['cirq.TransformerContext'] = None,
+    atol: float = 0.0,
+    eject_parameterized: bool = False,
+) -> 'cirq.Circuit':
+    """Pushes Z gates towards the end of the circuit.
+
+    As the Z gates get pushed they may absorb other Z gates, get absorbed into
+    measurements, cross CZ gates, cross PhasedXPowGate (aka W) gates (by phasing them), etc.
+
+    Args:
+          circuit: Input circuit to transform.
+          context: `cirq.TransformerContext` storing common configurable options for transformers.
+          atol: Maximum absolute error tolerance. The optimization is
+               permitted to simply drop negligible combinations of Z gates,
+               with a threshold determined by this tolerance.
+          eject_parameterized: If True, the optimization will attempt to eject
+              parameterized Z gates as well.  This may result in other gates
+              parameterized by symbolic expressions.
+    Returns:
+        Copy of the transformed input circuit.
+    """
+    # Tracks qubit phases (in half turns; multiply by pi to get radians).
+    qubit_phase: Dict[ops.Qid, float] = defaultdict(lambda: 0)
+    tags_to_ignore = set(context.tags_to_ignore) if context else set()
+    phased_xz_replacements: Dict[Tuple[int, ops.Operation], ops.PhasedXZGate] = {}
+    last_phased_xz_op: Dict[ops.Qid, Optional[Tuple[int, ops.Operation]]] = defaultdict(
+        lambda: None
+    )
+
+    def dump_tracked_phase(qubits: Iterable[ops.Qid]) -> 'cirq.OP_TREE':
+        """Zeroes qubit_phase entries by emitting Z gates."""
+        for q in qubits:
+            p, key = qubit_phase[q], last_phased_xz_op[q]
+            qubit_phase[q] = 0
+            if not (key or single_qubit_decompositions.is_negligible_turn(p, atol)):
+                yield ops.Z(q) ** (p * 2)
+            elif key:
```

Comment:
It looks like this key-checking block replaces the `if moment_index is not None` block in the old code, handling PhasedXZ tracking. Is that correct?

Reply:
Yes, that's correct. 

The idea here is that if the previous operation acting on the qubit `q` is a `PhasedXZGate` then instead of dumping a new `Z ** (2 * p)` gate on `q`, we simply modify the previous z exponent of the previous `PhasedXZGate`.